src/cdk-experimental/ui-patterns/tree/tree.ts

```diff
@@ -0,0 +1,519 @@
+  readonly keydown = computed(() => {
+    const manager = new KeyboardEventManager();
+
+    if (!this.followFocus()) {
+      manager
+        .on(this.prevKey, () => this.prev())
+        .on(this.nextKey, () => this.next())
+        .on('Home', () => this.first())
+        .on('End', () => this.last())
+        .on(this.typeaheadRegexp, e => this.search(e.key));
+    }
+
+    if (this.followFocus()) {
+      manager
+        .on(this.prevKey, () => this.prev({selectOne: true}))
+        .on(this.nextKey, () => this.next({selectOne: true}))
+        .on('Home', () => this.first({selectOne: true}))
+        .on('End', () => this.last({selectOne: true}))
+        .on(this.typeaheadRegexp, e => this.search(e.key, {selectOne: true}));
+    }
+
+    if (this.inputs.multi()) {
+      manager
+        .on(Modifier.Any, 'Shift', () => this.anchorItem.set(this.focusManager.activeItem()))
+        .on(Modifier.Shift, this.prevKey, () => this.prev({selectRange: true}))
+        .on(Modifier.Shift, this.nextKey, () => this.next({selectRange: true}))
+        .on([Modifier.Ctrl | Modifier.Shift, Modifier.Meta | Modifier.Shift], 'Home', () =>
+          this.first({selectRange: true, anchor: false}),
+        )
+        .on([Modifier.Ctrl | Modifier.Shift, Modifier.Meta | Modifier.Shift], 'End', () =>
+          this.last({selectRange: true, anchor: false}),
+        )
+        .on(Modifier.Shift, 'Enter', () =>
+          this._updateSelection({selectRange: true, anchor: false}),
+        )
+        .on(Modifier.Shift, this.dynamicSpaceKey, () =>
+          this._updateSelection({selectRange: true, anchor: false}),
+        );
+    }
+
+    if (!this.followFocus() && this.inputs.multi()) {
+      manager
+        .on(this.dynamicSpaceKey, () => this.selectionManager.toggle())
+        .on('Enter', () => this.selectionManager.toggle())
+        .on([Modifier.Ctrl, Modifier.Meta], 'A', () => this.selectionManager.toggleAll());
+    }
+
+    if (!this.followFocus() && !this.inputs.multi()) {
+      manager.on(this.dynamicSpaceKey, () => this.selectionManager.toggleOne());
+      manager.on('Enter', () => this.selectionManager.toggleOne());
+    }
+
+    if (this.inputs.multi() && this.followFocus()) {
+      manager
+        .on([Modifier.Ctrl, Modifier.Meta], this.prevKey, () => this.prev())
+        .on([Modifier.Ctrl, Modifier.Meta], this.nextKey, () => this.next())
+        .on([Modifier.Ctrl, Modifier.Meta], ' ', () => this.selectionManager.toggle())
+        .on([Modifier.Ctrl, Modifier.Meta], 'Enter', () => this.selectionManager.toggle())
+        .on([Modifier.Ctrl, Modifier.Meta], 'Home', () => this.first())
+        .on([Modifier.Ctrl, Modifier.Meta], 'End', () => this.last())
+        .on([Modifier.Ctrl, Modifier.Meta], 'A', () => {
+          this.selectionManager.toggleAll();
+          this.selectionManager.select(); // Ensure the currect item remains selected.
+        });
+    }
+
+    manager
+      .on(this.expandKey, () => this.expand())
```

Comment:
Based on my current understanding, if I navigate using the expandKey (arrow right on a parent treeitem that is already open) while holding shift, the range is not updated